File: solana/tests/support/burn-fixture.rs
// Synthetic finalized-burn evidence for program tests only. No chain proof or
// production mint authorization is conferred by this fixture.
use bridge_messages::{burn::{BurnBinding, BurnOutpoint, FinalizedBurnEvidence}, CanonicalBridgeMessage,
    DeploymentIdentity, MessageEpochs, NativeOutpoint, ValidityWindow};

pub struct TestBurnFields {
    pub deployment: DeploymentIdentity,
    pub deposit_outpoint: NativeOutpoint,
    pub amount_atomic: u64,
    pub solana_recipient: [u8;32],
    pub epochs: MessageEpochs,
    pub nonce: [u8;32],
    pub validity: ValidityWindow,
    pub evidence_digest: [u8;32],
}
pub fn burn_message(f: TestBurnFields) -> Result<CanonicalBridgeMessage,bridge_messages::MessageEncodeError> {
    let binding=BurnBinding {domain:*b"KPBOPR01",version:1,protocol_id:f.deployment.protocol_id,
        native_network:f.deployment.native_network,native_genesis:f.deployment.native_genesis,
        solana_genesis:if f.deployment.is_mainnet_bound(){bridge_messages::SOLANA_MAINNET_GENESIS}else{[242;32]},
        solana_deployment:f.deployment.solana_deployment,bridge_program:f.deployment.manager_program_id,
        transceiver_program:f.deployment.transceiver_program_id,mint:f.deployment.mint,
        destination:f.solana_recipient,burn_public_key:[241;32],nonce:f.nonce};
    let mut evidence=FinalizedBurnEvidence {domain:*b"KPBURN01",version:1,operation_id:binding.operation_id().unwrap(),binding,
        deposit:BurnOutpoint{txid:f.deposit_outpoint.txid,vout:f.deposit_outpoint.vout},deposit_block_hash:[243;32],deposit_height:10,
        burn:BurnOutpoint{txid:f.evidence_digest,vout:0},burn_block_hash:[244;32],burn_height:22,
        amount_atomic:f.amount_atomic,burn_commitment:[0;32]};
    evidence.burn_commitment=evidence.expected_commitment().unwrap();
    CanonicalBridgeMessage::new_deposit_claim(bridge_messages::DepositClaimFields {evidence,epochs:f.epochs,validity:f.validity})
}

#[allow(dead_code)]
pub fn rebind_burn_fixture(message: &mut CanonicalBridgeMessage) {
    let mut e=message.burn_evidence.clone();
    e.binding.protocol_id=message.deployment.protocol_id;
    e.binding.native_network=message.deployment.native_network;
    e.binding.native_genesis=message.deployment.native_genesis;
    e.binding.solana_deployment=message.deployment.solana_deployment;
    e.binding.bridge_program=message.deployment.manager_program_id;
    e.binding.transceiver_program=message.deployment.transceiver_program_id;
    e.binding.mint=message.deployment.mint;
    e.binding.destination=message.destination.clone().try_into().unwrap();
    e.binding.nonce=message.nonce;
    e.operation_id=e.binding.operation_id().unwrap();
    e.deposit=BurnOutpoint{txid:message.deposit_outpoint.txid,vout:message.deposit_outpoint.vout};
    e.amount_atomic=message.amount_atomic;
    e.burn_commitment=e.expected_commitment().unwrap();
    *message=CanonicalBridgeMessage::new_deposit_claim(bridge_messages::DepositClaimFields {
        evidence:e,epochs:MessageEpochs{policy_epoch:message.policy_epoch,key_epoch:message.key_epoch},
        validity:ValidityWindow{valid_from:message.valid_from,valid_until:message.valid_until}}).unwrap();
}

File: native/burn/tests/burn-fixture.mjs
// Copyright (c) 2026 KingPepe Team. All Rights Reserved.
// Synthetic cryptographic/parser fixture, never Native consensus evidence.
import {createHash} from 'node:crypto';
import {schnorr} from '@noble/curves/secp256k1.js';
import {ed25519} from '@noble/curves/ed25519.js';
import {deploymentFixture,mainnetDeploymentFixture} from '../../../tests/integration/deployment-fixture.mjs';
import {beginMainnetControlled} from '../../../services/bridge-validator/burn-mainnet-lifecycle.mjs';
import {base58Decode,base58Encode} from '../../../services/bridge-validator/solana-deposit-claim-transaction-plan.mjs';
import {decodeBridgeAbi,encodeBridgeAbi} from '../../../shared/protocol/solana-bridge-abi.mjs';
import {encodeBurnMessage} from '../../../shared/protocol/burn-message.mjs';
import {decodeCanonicalBridgeMessage} from '../../../shared/protocol/canonical-message.mjs';
import {burnOperationId,planNativeBurn,nativeBurnCommitment} from '../burn-protocol.mjs';
import {burnDepositDestination,burnOperationalDestination,signNativeBurnWithKey} from '../burn-key.mjs';
import {initialBurnJournal,issueBurnDeposit,recordBurnDeposits,retainBurnPlan,retainSignedBurn,markBurnBroadcast,retainFinalBurn} from '../../../services/bridge-validator/burn-journal-state.mjs';
import {ATTESTATION_PROTOCOL,ATTESTATION_MODE,BURN_SIGNED_BYTES} from '../../../services/attesters/burn-attestation-codec.mjs';
const hex=bytes=>Buffer.from(bytes).toString('hex'),hash=s=>createHash('sha256').update(s).digest('hex');
export function burnFixture({mainnet=false,amountAtomic='100000'}={}) {
  const {manifest,snapshot}=mainnet?mainnetDeploymentFixture(4):deploymentFixture(),root=schnorr.utils.randomSecretKey(),payer=ed25519.utils.randomSecretKey(),wallet=ed25519.utils.randomSecretKey(),attesters=[ed25519.utils.randomSecretKey(),ed25519.utils.randomSecretKey()];
  const pub=key=>hex(ed25519.getPublicKey(key));
  Object.assign(manifest.config,{protocolId:1,nativeNetwork:mainnet?manifest.config.nativeNetwork:8000111,attesters:attesters.map(a=>base58Encode(ed25519.getPublicKey(a)))});
  const tc=decodeBridgeAbi('TransceiverState',Buffer.from(snapshot.accounts[4].data[0],'base64'));
  Object.assign(tc.config,{protocolId:1,nativeNetwork:manifest.config.nativeNetwork,authorizedAttesters:attesters.map(seed=>ed25519.getPublicKey(seed))});
  snapshot.accounts[4].data[0]=encodeBridgeAbi('TransceiverState',tc).toString('base64');
  const binding={protocolId:1,nativeNetwork:manifest.config.nativeNetwork,nativeGenesis:manifest.nativeGenesisHex,solanaGenesis:hex(base58Decode(manifest.solanaGenesis)),solanaDeployment:manifest.solanaDeploymentHex,
    bridgeProgram:hex(base58Decode(manifest.manager.id)),transceiverProgram:hex(base58Decode(manifest.transceiver.id)),mint:hex(base58Decode(manifest.mint.id)),
    destination:pub(wallet),burnPublicKey:hex(schnorr.getPublicKey(root)),nonce:hash('fixture-nonce')};
  const {destination:_d,nonce:_n,...deployment}=binding,context={environment:manifest.environment,deployment,keyEpoch:1,policyEpoch:1,attesters:attesters.map(pub)};
  const artifact=Buffer.from(snapshot.accounts[5].data[0],'base64').subarray(45),entry={sha256:createHash('sha256').update(artifact).digest('hex'),byteLength:artifact.length};
  const policy={context,manifest,feePayerHex:pub(payer),artifacts:{manager:entry,transceiver:entry}};
  const state=initialBurnJournal(binding,{context,feePayerHex:pub(payer)});
  if(mainnet)beginMainnetControlled(state,{destinationHex:binding.destination,nonce:binding.nonce,amountModel:'EXACT_RECEIVED'});
  state.paused=false;state.pauseReason='TEST_REVIEWED';
  const op=issueBurnDeposit(state,binding,20),id=op.operationId,fees=burnOperationalDestination(binding);
  const deposit={txid:hash('deposit'),vout:0,amountAtomic,blockHash:hash('deposit-block'),height:21,confirmations:12};
  let plan=planNativeBurn({operationId:id,inputs:[{txid:deposit.txid,vout:0,amountAtomic:deposit.amountAtomic,scriptPubKeyHex:burnDepositDestination(binding).scriptPubKeyHex},
    {txid:hash('fee-coin'),vout:0,amountAtomic:'1000000',scriptPubKeyHex:fees.scriptPubKeyHex}],operationalScriptHex:fees.scriptPubKeyHex,
    feePolicy:{minimumRateAtomicPerKvB:'100',normalRateAtomicPerKvB:'1000',maximumRateAtomicPerKvB:'1000000',maximumAbsoluteFeeAtomic:'580000',dustRelayAtomicPerKvB:'3000'}});
  const evidence={binding,operationId:id,deposit:{txid:deposit.txid,vout:0},depositHeight:21,depositBlockHash:deposit.blockHash,burn:{txid:plan.txid,vout:0},
    burnHeight:33,burnBlockHash:hash('burn-block'),amountAtomic:deposit.amountAtomic,burnCommitment:nativeBurnCommitment({operationId:id,deposit:{txid:deposit.txid,vout:0},amountAtomic:deposit.amountAtomic})};
  if(mainnet)plan={...plan,transactionBlockHints:Object.fromEntries(plan.inputs.map(input=>[input.txid,deposit.blockHash]))};
  const finalize=()=>{recordBurnDeposits(state,id,[deposit]);retainBurnPlan(state,id,plan);retainSignedBurn(state,id,signNativeBurnWithKey({rootSecret:root,binding,plan}));markBurnBroadcast(state,id);retainFinalBurn(state,id,evidence);};
  const authorize=(validFrom='100',validUntil='3700')=>{
    const bytes=encodeBurnMessage({evidence,keyEpoch:1,policyEpoch:1,validFrom,validUntil}),m=decodeCanonicalBridgeMessage(bytes);
    return {encodedMessageHex:hex(bytes),attestations:attesters.map((seed,i)=>({protocol:ATTESTATION_PROTOCOL,mode:ATTESTATION_MODE,role:['ATTESTER_A','ATTESTER_B'][i],keyEpoch:1,policyEpoch:1,
      attesterPublicKeyHex:pub(seed),messageDigestHex:m.messageDigestHex,operationIdHex:burnOperationId(binding),signedBytes:BURN_SIGNED_BYTES,signatureHex:hex(ed25519.sign(bytes,seed)),state:'VERIFIED_READY'}))};
  };
  return {manifest,snapshot,policy,context,binding,state,id,plan,evidence,deposit,payer,root,attesters,finalize,authorize,
    destroy:()=>[root,payer,wallet,...attesters].forEach(k=>k.fill(0))};
}
